fix: deposit token allowance

xgambitox opened the pull request:
## Description

Fixes a race condition on vaults detail view, in which the wallet is not yet connected before it executes, making users unable to deposit.
![image](https://user-images.githubusercontent.com/18649057/166620410-29374ca3-f0a0-4c98-868a-76540f4eb006.png)

## Motivation and Context

Refactored the deposit modals to fetch token allowance directly instead of using useAllowance hook, since was unable to make it work through the hook.

## How Has This Been Tested?

Reloaded vault detail page multiple times with wallet connected.
Approved tokens in fantom network to test functionality. Also tested labs.
@FoxTheSin help needed with QA.

karelianpie commented on src/core/store/modules/vaults/vaults.actions.ts:
```diff
@@ -199,7 +198,7 @@ const approveDeposit = createAsyncThunk<
     });
 
-    await transactionService.handleTransaction({ tx: tx as TransactionResponse, network: network.current });
+    await transactionService.handleTransaction({ tx, network: network.current });
```
👍 for removing the casting

karelianpie commented on src/core/store/modules/vaults/vaults.actions.ts:
```diff
@@ -456,12 +455,22 @@ const getVaultAllowance = createAsyncThunk<
   if (!userAddress) throw new Error('WALLET NOT CONNECTED');
 
-  return vaultService.getVaultAllowance({
+  const tokenAllowance = await vaultService.getVaultAllowance({
```
I feel this sounds confusing, we are assigning the result of `getVaultAllowance()` to `tokenAllowance`?

xgambitox replied:
used that name since `TokenAllowance` is the type that is returned by this method. It is indeed a tokenAllowance. The service should be called `getVaultDepositAllowance` though imo. Since we will later add a different one for withdrawals.

karelianpie replied:
So not `getTokenAllowance()`?

xgambitox replied:
I would rename it to `getDepositAllowance` since its implicit it refers to vaults since its in the `vaultService`. That way its also aligned with sdk here https://github.com/yearn/yearn-sdk/blob/master/src/interfaces/vault.ts#L337. wdyt?

karelianpie replied:
That sounds good to me!

xgambitox replied:
refactored here :) https://github.com/yearn/yearn-finance-v3/pull/630/commits/e73ba66db30b3b4143aac516d62beae4ec40176b

karelianpie commented on src/client/components/app/Transactions/Labs/LabDepositTx.tsx:
```diff
@@ -149,9 +157,8 @@ export const LabDepositTx: FC<LabDepositTxProps> = ({ onClose }) => {
-      tokenAllowanceError,
-    loading: expectedTxOutcomeStatus.loading || isDebouncePending || isLoadingTokenAllowance,
+      slippageError,
+    loading: expectedTxOutcomeStatus.loading || isDebouncePending,
```
When `isFetchingAllowance` is not loading for the target status?

xgambitox replied:
it just disables the approve button when it is loading now, imo it should not affect the simulation output from showing, since either it is approved or not, but target expected outcome persists anyway

karelianpie commented on src/client/components/app/Transactions/Labs/LabDepositTx.tsx:
```diff
@@ -110,17 +121,14 @@ export const LabDepositTx: FC<LabDepositTxProps> = ({ onClose }) => {
-    tokenAmount: toBN(debouncedAmount),
-    tokenDecimals: selectedSellToken.decimals.toString(),
-    rawAllowance: tokenAllowance?.amount || '0',
+  const { approved: isApproved, error: allowanceError } = validateVaultAllowance({
```
How about here just changing the return from `validateVaultAllowance()` as it appears we were not and still aren't happy with the naming?

xgambitox replied:
not following the issue here. can you explain? not sure if we should rename the validation function name

karelianpie replied:
Sorry I wasn't very clear, what I meant was to change the returned object keys, so instead of `approve` we would return `isApproved` from the `validateVaultAllowance()` as it seems that's the name we want and it's best practice when returning a boolean to prefix with "is"

karelianpie replied:
Let's leave as is for now so we can resolve the bug and maybe then get back to this later. Resolving

karelianpie commented on src/client/components/app/Transactions/DepositTx.tsx:
```diff
@@ -54,8 +53,10 @@ export const DepositTx: FC<DepositTxProps> = ({
+  const [isFetchingAllowance, setIsFetchingAllowance] = useState(false);
   const currentNetwork = useAppSelector(NetworkSelectors.selectCurrentNetwork);
   const walletNetwork = useAppSelector(WalletSelectors.selectWalletNetwork);
+  const walletIsConnected = useAppSelector(WalletSelectors.selectWalletIsConnected);
```
nit: Usually the best practice is to just have these like below
```suggestion
  const isWalletConnected = useAppSelector(WalletSelectors.selectWalletIsConnected);
```
No action needed, but would be quicker to see that that returns a boolean

xgambitox replied:
Yeah, agree with this, just used the same naming from the selector, which probably should also be named as you describe

xgambitox replied:
changed here https://github.com/yearn/yearn-finance-v3/pull/630/commits/6f98c168328fc60e05d693fbaadc3c2a0a309621

karelianpie commented on src/client/components/app/Transactions/DepositTx.tsx:
```diff
@@ -109,6 +106,21 @@ export const DepositTx: FC<DepositTxProps> = ({
     };
   }, []);
 
+  useEffect(() => {
+    if (!selectedVault || !selectedSellTokenAddress || !walletIsConnected) return;
+    const fetchAllowance = async () => {
+      setIsFetchingAllowance(true);
+      await dispatch(
+        VaultsActions.getVaultAllowance({
+          vaultAddress: selectedVault.address,
+          tokenAddress: selectedSellTokenAddress,
+        })
+      );
+      setIsFetchingAllowance(false);
+    };
+    fetchAllowance();
+  }, [selectedSellTokenAddress, selectedVault?.address, walletIsConnected]);
```
Would it be possible to extract just this bit to its own hook so we can reuse in `LabDepositTx.tsx` -- if too hard let's just leave as is

xgambitox replied:
Yeah, it would be ideal. Should we refactor in a new PR, so we release a fix for this bug that is in prod now?

karelianpie replied:
Sounds good

FoxTheSin commented:
So I wasn't sure how to test this so I just did an approve in labs (not sure if relevant to this issue though)
(vbnt to yvBOOST)
Approval window was stuck open (despite tx going through)
![image](https://user-images.githubusercontent.com/79029523/167043296-77622b3e-8bf7-479a-b180-8fd0cfcfd6f9.png)

I then refreshed to try deposit, this happened
![image](https://user-images.githubusercontent.com/79029523/167043275-2a63bcac-dd2a-464b-abcd-9693144ae9b4.png)

xgambitox commented:
> So I wasn't sure how to test this so I just did an approve in labs (not sure if relevant to this issue though) (vbnt to yvBOOST) Approval window was stuck open (despite tx going through) ![image](https://user-images.githubusercontent.com/79029523/167043296-77622b3e-8bf7-479a-b180-8fd0cfcfd6f9.png)
> 
> refreshed page after confirming "approval", and before clicking deposit, simulation appeared to pass ![image](https://user-images.githubusercontent.com/79029523/167043445-226c0046-9c34-4390-a5ad-7ae2d920add4.png)
> 
> I then I clicked deposit and this happened ![image](https://user-images.githubusercontent.com/79029523/167043275-2a63bcac-dd2a-464b-abcd-9693144ae9b4.png)

the transaction notification never works on preview builds since the generated URL is not whitelisted in blocknatives dashboard.

About the error from vBNT -> yvboost, not sure if its related to this PR. Can you try in prod if same thing happens? That is how we can discard it from being related to this one.

I would do QA with more liquid tokens though, since we might get other type of error non related to this PR if not.

FoxTheSin commented:
> 
> the transaction notification never works on preview builds since the generated URL is not whitelisted in blocknatives dashboard.
> 
> About the error from vBNT -> yvboost, not sure if its related to this PR. Can you try in prod if same thing happens? That is how we can discard it from being related to this one.
> 
> I would recommend doing QA with more liquid tokens though, since we might get other type of error non related to this PR if not.

Yup, doesn't work in prod.
but everything else looked good!

FoxTheSin commented:
another weird occurrence, 
1) opened the amplify for pr (wallet was set to eth)
2) switched to ftm from meta mask (in portfolio page)
3) changed to vaults page
4) see below! (list of eth and ftm got combined.)

![image](https://user-images.githubusercontent.com/79029523/167044899-fd706063-423e-4dcb-97fd-8015b80c20b0.png)

xgambitox commented:
> another weird occurrence (but unable to reproduce)
> 
> 1. opened the amplify for pr (wallet was set to eth)
> 2. switched to ftm from meta mask (in portfolio page)
> 3. changed to vaults page
> 4. see below! (list of eth and ftm got combined.)
> 
> ![image](https://user-images.githubusercontent.com/79029523/167044899-fd706063-423e-4dcb-97fd-8015b80c20b0.png)

Yeah, I have seen this one, and might be good to track on another issue. It seems related to fetching been slow, and therefore vaults state not getting cleared properly

karelianpie reviewed (approved):
LGTM

turtlemoji commented:
@xgambitox maybe bump package so we can just release

xgambitox commented:
@FoxTheSin @karelianpie @turtlemoji should be ready now 👍
